proof/refine/ARM/InterruptAcc_R.thy

```diff
@@ -93,28 +93,357 @@ lemma work_units_and_irq_state_state_relationI [intro!]:
    \<in> state_relation"
   by (simp add: state_relation_def swp_def)
 
+lemma update_work_units_corres[corres]:
+  "corres (dc \<oplus> dc) \<top> \<top> (liftE update_work_units) (liftE (modifyWorkUnits (\<lambda>op. op + 1)))"
+  apply (clarsimp simp: update_work_units_def modifyWorkUnits_def)
+  apply (rule corres_modify)
+  apply (clarsimp simp: state_relation_def)
+  done
+
+lemma getCurTime_corres[corres]:
+  "corres (=) \<top> \<top> (gets cur_time) getCurTime"
+  apply (simp add: getCurTime_def state_relation_def)
+  done
+
+lemma getCurTime_sp:
+  "\<lbrace>P\<rbrace> getCurTime \<lbrace>\<lambda>rv s. rv = ksCurTime s \<and> P s\<rbrace>"
+  apply wpsimp
+  done
+
+lemma updateTimeStamp_corres[corres]:
+  "corres dc \<top> \<top> update_time_stamp updateTimeStamp"
+  apply (clarsimp simp: update_time_stamp_def updateTimeStamp_def setConsumedTime_def)
+  apply (rule corres_split'[rotated 2, OF gets_sp getCurTime_sp])
+   apply corressimp
+  apply (rule corres_split'[where r'="(=)"])
+     apply (rule corres_guard_imp)
+       apply (rule corres_machine_op)
+       apply corressimp
+        apply (wpsimp simp: getCurrentTime_def)
+       apply simp
+      apply simp
+     apply simp
+    apply (rule_tac P=\<top> and P'=\<top> in corres_inst)
+    apply (clarsimp simp: setCurTime_def)
+    apply (rule corres_guard_imp)
+      apply (rule corres_split[OF corres_modify])
+         apply (clarsimp simp: state_relation_def cdt_relation_def)
+        apply (clarsimp simp: setConsumedTime_def)
+        apply (rule_tac Q'="\<lambda>rv s. rv = ksConsumedTime s" in corres_symb_exec_r)
+           apply (rule corres_modify)
+           apply (clarsimp simp: state_relation_def cdt_relation_def)
+          apply (wpsimp simp: getConsumedTime_def)+
+  done
+
+lemma sc_at'_cross_rel:
+  "cross_rel (pspace_aligned and pspace_distinct and sc_at t) (sc_at' t)"
```

Comment:
Maybe move? Or comment to that effect?

This is actually weaker than `sc_at_cross`, but I guess you specifically wanted `cross_rel`.  Anyway, `sc_at_cross` is in KHeap_R, so maybe that’s where this can go, too.